src/emoji/data.js

```diff
@@ -28,12 +28,34 @@ export const getFilteredEmojis = (
   query: string,
   activeImageEmojiByName: $ReadOnly<{ [string]: ImageEmojiType }>,
 ): $ReadOnlyArray<{| emoji_type: EmojiType, name: string, code: string |}> => {
+  const matchingUnicodeEmoji = Object.entries(unicodeCodeByName).map(x => {
+    if (typeof x[1] !== 'string') {
+      // This is impossible, but flow isn't smart enough to figure that out.
+      throw new Error('Emoji name is not a string');
+    }
+    // We may want to have fancier emoji-matching logic in the future: this
+    // code will not work with skin-tone modifiers, for instance, and
+    // there may be other strange unicode things that warrant special
+    // handling. It would be simple to add code to strip off the
+    // skin-tone modifiers in the query, but it seems worse to suggest to
+    // the user the default skin color emoji when they've typed in a
+    // modified one than to display nothing. We should circle back to
+    // this later, and figure out what our overall strategy for skin-tone
+    // modifiers across the app should be.
+    const matches_emoji_literal = parseUnicodeEmojiCode(x[1]) === query;
+    const matches_emoji_name = Math.min(1, x[0].indexOf(query));
+    return [x[0], matches_emoji_literal ? 0 : matches_emoji_name];
+  });
+
+  const matchingCustomEmoji = Object.keys(activeImageEmojiByName).map(x => [
+    x,
+    Math.min(1, x.indexOf(query)),
+  ]);
+
   // Map from emoji names to 0 for prefix matches, 1 for others.
   // (Nonmatching names are excluded.)
   const nameMap: Map<string, number> = new Map(
-    [...unicodeEmojiNames, ...Object.keys(activeImageEmojiByName)]
-      .map(x => [x, Math.min(1, x.indexOf(query))])
-      .filter(([_, i]) => i !== -1),
+    [...matchingUnicodeEmoji, ...matchingCustomEmoji].filter(([_, i]) => i !== -1),
```

Comment:
The names with "matching" are a bit misleading: these lists actually contain all the emoji, just with markers indicating whether they match (and with what priority if so.)

One way to fix that would be to propagate the `filter` to the two definitions. That should also be slightly more efficient, as we then don't allocate a long array here if there are only a few matches.